src/cmap/background.rs

```diff
@@ -24,27 +27,46 @@ pub(crate) fn start_background_task(pool: Weak<ConnectionPoolInner>) {
 async fn perform_checks(pool: ConnectionPool) {
     // We remove the perished connections first to ensure that the number of connections does not
     // dip under the min pool size due to the removals.
-    pool.remove_perished_connections().await;
-    pool.ensure_min_connections().await;
+    pool.inner.remove_perished_connections().await;
+    pool.inner.ensure_min_connections().await;
 }
 
-impl ConnectionPool {
+impl ConnectionPoolInner {
     /// Iterate over the connections and remove any that are stale or idle.
     async fn remove_perished_connections(&self) {
-        let mut connection_manager = self.inner.connection_manager.lock().await;
+        // re-acquire the lock between loop iterations to allow other threads to use the pool.
+        loop {
+            let mut available_connections = self.available_connections.lock().await;
+
+            if available_connections.len() == 0 {
+                break;
+            }
 
-        let mut i = 0;
-        while i < connection_manager.checked_in_connections.len() {
-            if connection_manager.checked_in_connections[i].is_stale(connection_manager.generation)
-            {
-                let connection = connection_manager.checked_in_connections.remove(i);
-                connection_manager.close_connection(connection, ConnectionClosedReason::Stale);
-            } else if connection_manager.checked_in_connections[i].is_idle(self.inner.max_idle_time)
-            {
-                let connection = connection_manager.checked_in_connections.remove(i);
-                connection_manager.close_connection(connection, ConnectionClosedReason::Idle);
+            let connection = &available_connections[0];
+
+            let close_request = if connection.is_stale(self.generation.load(Ordering::SeqCst)) {
+                Some((
+                    available_connections.remove(0),
+                    ConnectionClosedReason::Stale,
+                ))
+            } else if connection.is_idle(self.max_idle_time) {
+                Some((
+                    available_connections.remove(0),
+                    ConnectionClosedReason::Idle,
+                ))
             } else {
-                i += 1;
+                None
```

Comment:
I think you could simplify this by breaking here; then you wouldn't need `close_request` to be an Option.

Reply:
done